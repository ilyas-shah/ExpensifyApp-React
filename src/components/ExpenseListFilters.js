// import 'react-dates/initialize';
import React from 'react';
import {connect} from 'react-redux';
import { DateRangePicker } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';

 export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            calendarFocused: null,
            focusedInput: true
        }
    }
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (calendarFocused) => {
        this.setState({calendarFocused})
    }
    onSortChange = (e) => {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount()
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    render() {
        return (
            <div>
            <input
                type="text"
                value={this.props.filters.text}
                onChange={this.onTextChange}
            />

            <select
                value = {this.props.filters.sortBy} 
                onChange={this.onSortChange}
            >
                <option value="date">Date</option>
                <option value="amount">Amount</option>
            </select>
            <DateRangePicker
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() =>(false)}
                startDate={this.props.filters.startDate}
                startDateId="313"
                endDate={this.props.filters.endDate}
                endDateId="312"
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
            />
        </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => (dispatch(setTextFilter(text))),
        setStartDate: (startDate) => (dispatch(setStartDate(startDate))),
        setEndDate: (endDate) => (dispatch(setEndDate(endDate))),
        sortByDate: () => (dispatch(sortByDate())),
        sortByAmount: () => (dispatch(sortByAmount()))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
