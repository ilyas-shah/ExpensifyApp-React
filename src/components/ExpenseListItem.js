import React from 'react';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ({description, amount, createdAt, id}) => (
    <div>
        <Link to={`/edit/${id}`}><h2>{description}</h2></Link>
        <p>{amount} - {createdAt}</p>
    </div>
)

export default ExpenseListItem;