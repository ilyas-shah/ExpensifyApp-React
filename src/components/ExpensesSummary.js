import numeral from 'numeral';
import React from 'react';
import { connect } from 'react-redux';
import selectedExpenses from '../selectors/expenses';
import expenseDetails from '../selectors/expensesTotal';

export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses'
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00')
    return (
        <div>
            {expensesCount > 0 && <p>{`viewing ${expensesCount} ${expenseWord} and totalling ${formattedExpensesTotal}`}</p> }
        </div>
    )
};

const mapStateToProps = (state) => {
  return {
    expensesCount: selectedExpenses(state.expenses, state.filters).length,
    expensesTotal: expenseDetails(selectedExpenses(state.expenses, state.filters))
  }
}

export default connect(mapStateToProps)(ExpensesSummary);
