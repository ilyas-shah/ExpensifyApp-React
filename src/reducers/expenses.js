// Expenses Reducer

const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense]
        break;
    
    case 'REMOVE_EXPENSE':
      return state.filter(({id}) => 
        id !== action.id
      )
      break;
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (action.id === expense.id) {
          return {
            ...expense,
            ...action.updates
          }
        } else {
          return {
            ...expense 
          }
        }
      })
      break;
    default:
      return state;
  }
};

export default expensesReducer;