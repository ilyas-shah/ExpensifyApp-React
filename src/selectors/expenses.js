import moment from 'moment';
// VISIBLE EXPENSES
const getVisibleExpenses = (expenses, { startDate, endDate, text, sortBy }) => {
    let startDateMatch
    let endDateMatch
    let textMatch
  
    return expenses.filter((expense) => {
      const createdAtMoment = moment(expense.createdAt)
      startDateMatch = startDate ? startDate.isSameOrAfter(createdAtMoment, 'day') : true;
      endDateMatch = endDate ? endDate.isSameOrBefore(createdAtMoment, 'day') : true;
      textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
  
      return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1
      } else {
        return 0
      }
    })
}
export default getVisibleExpenses;