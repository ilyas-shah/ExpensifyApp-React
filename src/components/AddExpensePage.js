import React from 'react';
import ExpenseForm from './ExpenseForm';
import expenses from '../reducers/expenses';
import {connect} from 'react-redux';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.onSubmit(expense)
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
        <h2>Add Expense</h2>
        <ExpenseForm
          onSubmit={ this.onSubmit }
        />
      </div>
    )
  }
}

// const AddExpensePage = (props) => (
//   <div>
//     <h2>Add Expense</h2>
//     <ExpenseForm
//       onSubmit={(expense) => {
//         props.dispatch(addExpense(expense))
//         props.history.push('/')
//       }}
//     />
//   </div>
// );

const mapDispatchToProps = (dispatch) => ({ 
  onSubmit: (expense) => (dispatch(addExpense(expense)))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
