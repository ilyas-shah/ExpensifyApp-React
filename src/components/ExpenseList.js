import React from 'react';
import {connect} from 'react-redux';
import expenses from '../selectors/expenses';
import ExpenseListItem from '../components/ExpenseListItem';
import selectedExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ?
            (
                <h2>No Expense</h2>
            )
            : 
            props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense}/>
            })
        }
    </div>
)
const mapStateToProps = (state, context) => {
    return {
        expenses: selectedExpenses(state.expenses, state.filters)
    }
}
export default connect(mapStateToProps)(ExpenseList);