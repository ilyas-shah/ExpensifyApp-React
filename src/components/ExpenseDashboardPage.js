import React from 'react';
import ExpenseList from '../components/ExpenseList';
import ExpenseListFilters from '../components/ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
