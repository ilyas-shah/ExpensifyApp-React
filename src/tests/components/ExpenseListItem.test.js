import React from 'react';
import { shallow } from 'enzyme';
import toJSON from "enzyme-to-json";
import { expenses }  from '../../fixtures/expenses';
import { ExpenseListItem } from '../../components/ExpenseListItem';

test('should render Expense list item correctly', () => {
    const expense = expenses[0]
    const wrapper = shallow(<ExpenseListItem {...expense}/>)
    expect(toJSON(wrapper)).toMatchSnapshot()
});