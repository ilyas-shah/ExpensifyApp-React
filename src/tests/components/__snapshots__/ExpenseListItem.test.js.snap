// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense list item correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/123456790"
  >
    <h2>
      Gum
    </h2>
  </Link>
  <p>
    10
     - 
    0
  </p>
</div>
`;
